TTS/tts/models/forward_tts_e2e.py

```diff
@@ -0,0 +1,1002 @@
+        encoder_outputs = self.encoder_model.inference(x=x, aux_input=aux_input, skip_decoder=False)
+        model_outputs = {**encoder_outputs}
+        return model_outputs
+
+    @staticmethod
+    def init_from_config(config: "ForwardTTSConfig", samples: Union[List[List], List[Dict]] = None, verbose=False):
+        """Initiate model from config
+
+        Args:
+            config (ForwardTTSE2eConfig): Model config.
+            samples (Union[List[List], List[Dict]]): Training samples to parse speaker ids for training.
+                Defaults to None.
+        """
+        from TTS.utils.audio.processor import AudioProcessor
+
+        tokenizer, new_config = TTSTokenizer.init_from_config(config)
+        speaker_manager = SpeakerManager.init_from_config(config, samples)
+        # language_manager = LanguageManager.init_from_config(config)
+        return ForwardTTSE2e(config=new_config, tokenizer=tokenizer, speaker_manager=speaker_manager)
+
+    def load_checkpoint(
+        self, config, checkpoint_path, eval=False
+    ):  # pylint: disable=unused-argument, redefined-builtin
+        state = torch.load(checkpoint_path, map_location=torch.device("cpu"))
+        self.load_state_dict(state["model"])
+        if eval:
+            self.eval()
+            assert not self.training
+
+    def train_step(self, batch: dict, criterion: nn.Module, optimizer_idx: int):
+        if optimizer_idx == 0:
+            tokens = batch["text_input"]
+            token_lenghts = batch["text_lengths"]
+            spec = batch["mel_input"]
+            spec_lens = batch["mel_lengths"]
+            waveform = batch["waveform"]  # [B, T, C] -> [B, C, T]
+            pitch = batch["pitch"]
+            d_vectors = batch["d_vectors"]
+            speaker_ids = batch["speaker_ids"]
+            language_ids = batch["language_ids"]
+
+            # generator pass
+            outputs = self.forward(
+                x=tokens,
+                x_lengths=token_lenghts,
+                spec_lengths=spec_lens,
+                spec=spec,
+                waveform=waveform,
+                pitch=pitch,
+                aux_input={"d_vectors": d_vectors, "speaker_ids": speaker_ids, "language_ids": language_ids},
+            )
+
+            # cache tensors for the generator pass
+            self.model_outputs_cache = outputs  # pylint: disable=attribute-defined-outside-init
+
+            # compute scores and features
+            scores_d_fake, _, scores_d_real, _ = self.disc(outputs["model_outputs"].detach(), outputs["waveform_seg"])
+
+            # compute loss
+            with autocast(enabled=False):  # use float32 for the criterion
+                loss_dict = criterion[optimizer_idx](
+                    scores_disc_fake=scores_d_fake,
+                    scores_disc_real=scores_d_real,
+                )
+            return outputs, loss_dict
+
+        if optimizer_idx == 1:
+            mel = batch["mel_input"].transpose(1, 2)
```

Comment:
Maybe implementing `steps_to_start_discriminator` would allow for faster training

Reply:
Good idea! I'll try.